### PR TITLE
Port p3 main

### PR DESCRIPTION
**Pull request by jgfouca**
This is a big day for SCREAM. With this PR, all P3 code that we care about is ported. There will probably still be follow-on PRs to cleanup things.

Other change list:
* Added 2D version of host_to_device and device_to_host

### CODE REVIEW

**Comment by ambrad** [comment]
Is p3_run_and_cmp now calling the C++ code at the highest level possible? Best would be calling it directly, with no F90 intermediary.

File: components/scream/ekat/tests/pack/pack_kokkos_tests.cpp [hunk 1]
```diff
@@ -299,13 +299,13 @@ TEST_CASE("host_device_packs", "scream::pack")
   Kokkos::Array<size_t, num_views_per_pksize> sizes = {13, 37, 59}; // num scalars per view
   std::vector<std::vector<int> > raw_data(num_pksizes_to_test, std::vector<int>());
 
-  int total_scalars = 0; // each pksize has total_scalars of data
+  int total_flex_scalars = 0; // each pksize has total_flex_scalars of data
```
Comment (jeff-cohere) [inline components/scream/ekat/tests/pack/pack_kokkos_tests.cpp:302]: Is it worth mentioning what "flex" or "flexible" means in this context? Is this a new concept, or do we have this elsewhere, and is it clear what it is?

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 2]
```diff
@@ -220,11 +228,13 @@ namespace p3 {
...
 namespace {
 
-template <size_t N>
+template <size_t N, size_t M>
 void gen_random_data(const std::array<std::pair<Real, Real>, N>& ranges,
-                     const std::array<Real**, N>& ptrs,
+                     const std::array<Real**, M>& ptrs,
                      Real* data, Int nk)
 {
+  static_assert(N <= M);
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_f90.cpp:236]: This is syntactically wrong; static_assert requires a message as a second arg. I'm curious why the GPU build isn't seeing that. Is this used only in CPU builds?

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 3]
```diff
@@ -2750,26 +2814,8 @@ void get_latent_heat_f(Int its, Int ite, Int kts, Int kte, Real* v, Real* s, Rea
...
-  std::copy(v, v+total, temp.v);
-  std::copy(s, s+total, temp.s);
-  std::copy(f, f+total, temp.f);
-
-  temp.transpose();
-
-  std::copy(temp.v, temp.v+total, v);
-  std::copy(temp.s, temp.s+total, s);
-  std::copy(temp.f, temp.f+total, f);
+  Kokkos::Array<view_2d, 3> out_views = {v_d, s_d, f_d};
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_f90.cpp:2817]: Nice replacement. @pressel, you might look at to guide your SHOC work; you can use the new utils that Jim has created by merging this PR into your branch. (Jim, am I correct that you won't rebase this branch, so it's safe to merge it into other branches?)

File: components/scream/src/physics/p3/p3_functions_f90.cpp (hunk 2 above)
Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:236]: Apparently this construction is allowed in C++17:

https://en.cppreference.com/w/cpp/language/static_assert

Though I don't know that this explains why it compiled with our settings.

Comment (ambrad) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:236]: Oh, nice! Can't wait until 2017.

File: components/scream/src/physics/p3/p3_functions_f90.cpp (hunk 3 above)
Comment (jgfouca) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:2817]: Yes, I will not rebase. I may do upstream merge though.

**Comment by jgfouca** [comment]
@ambrad , yes, follow-on PRs will take advantage of the completed conversion to simply p3_run_and_cmp and micro_p3.

File: components/scream/src/physics/p3/p3_functions_f90.cpp (hunk 2 above)
Comment (jgfouca) [reply components/scream/src/physics/p3/p3_functions_f90.cpp:236]: Looks like only bowman caught this. Fixing now.

File: components/scream/ekat/tests/pack/pack_kokkos_tests.cpp (hunk 1 above)
Comment (jgfouca) [reply components/scream/ekat/tests/pack/pack_kokkos_tests.cpp:302]: It's just to distinguish from the fixed-size views which are also tested here. I'll try to improve these comments.

**Comment by jgfouca** [comment]
I think everyone's feedback has been addressed. I will do a follow-on PR with framework cleanup.

**Pull request merged**